### PR TITLE
chore: Update custom-renderer codelab to modern setup.

### LINKED ISSUE #1545: Update custom renderer codelab to use sample app, and include example code
Following the pattern laid down in https://github.com/google/blockly-samples/pull/1439

### PR DESCRIPTION
**Pull request by jubishop**
This change attempts to follow the structure, writing style and instructions in the custom-generator codelab and reproduce it in the custom-renderer codelab.

Some things of note:

- The "What you'll learn" items are now dashes instead of numbered.
- All embedded images now have accessibility alt descriptions.
- The es5 file and syntax is all gone and updated to es6.
- We start with the sample-app instead of the playground.html
- The user now creates a new `src/renderers` folder and adds a `custom.js`.
- Some links to source files were broken because things had been moved or files now end in `.ts` instead of `.js`
- In the places where we include a snippet of code from production blockly, the snippet was no longer a match to what actually is in our code, so I updated these.
  - I also made sure to link to the full code inside the `blockly` github repo.

fixes: #1545

### CODE REVIEW

File: codelabs/custom_renderer/custom_renderer.md [hunk 1]
```diff
@@ -1,143 +1,176 @@
...
+// Use a unique storage key for this codelab
+const storageKey = 'customRenderersWorkspace';
+```
+
+## Define and register a custom renderer
 
 A **Renderer** is the interface between your custom rendering code and the rest of Blockly. Blockly provides a base renderer with all required fields set to default values.
 
-Your new renderer must extend the base renderer:
+To start, create a new directory at `src/renderers` and add a file inside named `javascript.js`.
```
Comment (rachel-fenichel) [inline codelabs/custom_renderer/custom_renderer.md]: Why is the renderer file `javascript.js` instead of `custom.js` or similar?

File: codelabs/custom_renderer/custom_renderer.md [hunk 2]
```diff
@@ -1,143 +1,176 @@
...
-    renderer: 'custom_renderer'
-  }
-);
+import './renderers/javascript';
+```
+
+Now, add the the `renderer` property into the configuration struct passed to `Blockly.inject` so that it now looks like this:
+
+```js
+const ws = Blockly.inject(blocklyDiv, {renderer: 'custom_renderer', toolbox});
```
Comment (rachel-fenichel) [inline codelabs/custom_renderer/custom_renderer.md]: If possible, split into multiple lines to make the `renderer: 'custom_renderer'` part stand out from the `toolbox` and `blocklyDiv`

File: codelabs/custom_renderer/custom_renderer.md [hunk 3]
```diff
@@ -1,143 +1,176 @@
...
 ```
 
 ### The result
 
-To test, open the playground in your browser and drag out a repeat block. The resulting block looks close to normal, but with slightly different alignment than Geras or Thrasos.
+If the server is already running, you can refresh the page to see your changes. Otherwise, run `npm run start` to start the server. Once the server is running, click on the `Loops` entry in your browser and drag out a repeat block. The resulting block will look the same as normal, since your custom renderer has merely extended the default renderer without changing anything.
```
Comment (rachel-fenichel) [inline codelabs/custom_renderer/custom_renderer.md]: I expect the alignment to still be off (as described in the previous text), because this extends the base renderer (`Blockly.blockRendering.Renderer`) instead of the default. Can you double-check?

File: codelabs/custom_renderer/custom_renderer.md [hunk 4]
```diff
@@ -1,143 +1,176 @@
...
+A **ConstantsProvider** holds all rendering-related constants.  This includes sizing information and colors. Blockly provides a base **ConstantsProvider** with all required fields set to default values.
 
-The **ConstantsProvider** constructor sets all static properties, such as `NOTCH_WIDTH` and `NOTCH_HEIGHT`. For a full list of properties, see [constants.js](https://github.com/google/blockly/blob/master/core/renderers/common/constants.js).
+The **ConstantsProvider** `constructor()` sets all static properties, such as `NOTCH_WIDTH` and `NOTCH_HEIGHT`. For a full list of properties, see [constants.ts](https://github.com/google/blockly/blob/master/core/renderers/common/constants.ts).
 
 In general you will want to override a subset of the constants, rather than all of them. To do so:
-- Define a constants provider that extends the base provider.
-- Call the superclass constructor in your constructor.
+- Define a constants provider that extends the base `ConstantProvider`.
+- Call the superclass `constructor()` in your `constructor()`.
```
Comment (rachel-fenichel) [inline codelabs/custom_renderer/custom_renderer.md]: Suggestion: "Call `super()` in your constructor.

File: codelabs/custom_renderer/custom_renderer.md [hunk 5]
```diff
@@ -147,221 +180,228 @@ A common use case of a custom renderer is changing the shape of connections. Thi
...
- * @param {!Blockly.RenderedConnection} connection The connection to find a
- *     shape object for
- * @returns {!Object} The shape object for the connection.
- * @package
+ * Get an object with connection shape and sizing information based on the
+ * type of the connection.
+ *
+ * @param connection The connection to find a shape object for
+ * @returns The shape object for the connection.
+ * @internal
```
Comment (rachel-fenichel) [inline codelabs/custom_renderer/custom_renderer.md]: @maribethb is this going to be public instead of internal after the next release?

Comment (rachel-fenichel) [reply codelabs/custom_renderer/custom_renderer.md]: Same question for any other internal function.

File: examples/custom-renderer-codelab/package.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "renderer-sample-app",
+  "version": "1.0.0",
+  "description": "A sample app using Blockly",
```
Comment (rachel-fenichel) [inline examples/custom-renderer-codelab/package.json]: Update description to connect to the codelab somehow.

File: examples/custom-renderer-codelab/src/renderers/javascript.js [hunk 7]
```diff
@@ -1,16 +1,6 @@
...
- */
-
-/**
- * @fileoverview The full custom renderer built during the custom renderer
- * codelab, in ES6 syntax.
- * @author fenichel@google.com (Rachel Fenichel)
- */
-
-class CustomConstantsProvider extends Blockly.blockRendering.ConstantProvider {
+import * as Blockly from 'blockly/core';
```
Comment (rachel-fenichel) [inline examples/custom-renderer-codelab/src/renderers/javascript.js]: Add license and fileoverview back. Author tag is optional now.

**Review by rachel-fenichel — CHANGES_REQUESTED** [review]

File: codelabs/custom_renderer/custom_renderer.md (hunk 5 above)
Comment (maribethb) [reply codelabs/custom_renderer/custom_renderer.md]: Yes, none of these methods should be annotated with `@internal` anymore (for maximum clarity, we don't use the `@public` annotation, there just should be no annotation now)

File: codelabs/custom_renderer/custom_renderer.md [hunk 8]
```diff
@@ -1,143 +1,176 @@
...
-To test, open the playground in your browser and drag out a repeat block. The resulting block looks close to normal, but with slightly different alignment than Geras or Thrasos.
+If the server is already running, you can refresh the page to see your changes. Otherwise, run `npm run start` to start the server. Once the server is running, click on the `Loops` entry in your browser and drag out a repeat block. The resulting block will look the same as normal, since your custom renderer has merely extended the default renderer without changing anything.
 
-![](./custom_renderer.png)
+![Screenshot of a renderer with a default appearance.](./custom_renderer.png)
 
 ## Override constants
 
-A **ConstantsProvider** holds all rendering-related constants.  This includes sizing information and colours.  Blockly provides a base **ConstantsProvider** with all required fields set to default values.
+A **ConstantsProvider** holds all rendering-related constants.  This includes sizing information and colors. Blockly provides a base **ConstantsProvider** with all required fields set to default values.
```
Comment (rachel-fenichel) [inline codelabs/custom_renderer/custom_renderer.md]: Please use `colour` instead of `color`, so that the docs match the code.

But here's a video of [eagles harassing a beaver](https://www.youtube.com/watch?v=8xw6AxbtSi8) to make up for losing that iota of life, liberty, and the pursuit of happiness.

File: codelabs/custom_renderer/custom_renderer.md (hunk 1 above)
Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: this was in the commit description.  The user now creates a new src/renderers folder and adds a javascript.js which seems to match the structure of the rest of the folders and files already in the sample app such as src/generators/javascript.js.   

so basically just matching how it looked in src/generators.  wdyt?

Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: moved to custom.js

File: examples/custom-renderer-codelab/src/renderers/javascript.js (hunk 7 above)
Comment (jubishop) [reply examples/custom-renderer-codelab/src/renderers/javascript.js]: should https://github.com/google/blockly-samples/blob/master/examples/custom-generator-codelab/src/generators/json.js and https://github.com/google/blockly-samples/blob/master/examples/custom-generator-codelab/src/blocks/json.js have fileoverview's too?

Comment (jubishop) [reply examples/custom-renderer-codelab/src/renderers/javascript.js]: i've updated this one. leaving unresolved to find out about my question (i will put a fileoverview on those, if so)

File: examples/custom-renderer-codelab/package.json (hunk 6 above)
Comment (jubishop) [reply examples/custom-renderer-codelab/package.json]: yeah. this was on my todo list to change after maribeth's review on https://github.com/google/blockly-samples/pull/1601, just hadn't gotten to it yet.  good catch!

File: codelabs/custom_renderer/custom_renderer.md (hunk 8 above)
Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: i guess the brits win this round.  *shakes fist*

Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: oh and also, Done.

File: codelabs/custom_renderer/custom_renderer.md (hunk 2 above)
Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: cool.  i'm not experienced with how this sort of stuff should be formatted, but i changed it to something that the linter seems to be happy with and puts it on its own line.  also updated in the `index.js` file.

File: codelabs/custom_renderer/custom_renderer.md (hunk 4 above)
Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: good point, that makes a lot more sense.  done.

File: examples/custom-renderer-codelab/src/renderers/javascript.js (hunk 7 above)
Comment (jubishop) [reply examples/custom-renderer-codelab/src/renderers/javascript.js]: (in a separate PR, to be clear!)

File: codelabs/custom_renderer/custom_renderer.md (hunk 3 above)
Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: ohh. what's the default?  (Geras or Thrasos?).  when i was reading through this those words made no sense to me.  nowhere else is "Geras" or "Thrasos" mentioned so I was like, wtf is that?

File: codelabs/custom_renderer/custom_renderer.md [hunk 9]
```diff
@@ -1,143 +1,176 @@
...
-![](./custom_renderer.png)
-1. A renderer with custom constants.
-![](./custom_constants.png)
-1. A renderer with custom notches.
-![](./custom_notches.png)
-1. A renderer with connection shapes that depend on the type checks of each connection.
-![](./typed_connection_shapes.png)
-
-The code samples are written in ES5 syntax. You can find the code for the [completed custom renderer](https://github.com/google/blockly-samples/tree/master/codelabs/custom_renderer) on GitHub, in both ES5 and ES6 syntax.
+1. A minimal custom renderer that extends `Blockly.blockRendering.Renderer` but makes no modifications to the default rendering style.
```
Comment (jubishop) [inline codelabs/custom_renderer/custom_renderer.md]: depending on what we decide for how to describe this in the other comment review, i'll want to update this too.

File: codelabs/custom_renderer/custom_renderer.md [hunk 10]
```diff
@@ -1,143 +1,176 @@
...
+const ws = Blockly.inject(blocklyDiv, {renderer: 'custom_renderer', toolbox});
 ```
 
 ### The result
 
-To test, open the playground in your browser and drag out a repeat block. The resulting block looks close to normal, but with slightly different alignment than Geras or Thrasos.
+If the server is already running, you can refresh the page to see your changes. Otherwise, run `npm run start` to start the server. Once the server is running, click on the `Loops` entry in your browser and drag out a repeat block. The resulting block will look the same as normal, since your custom renderer has merely extended the default renderer without changing anything.
 
-![](./custom_renderer.png)
+![Screenshot of a renderer with a default appearance.](./custom_renderer.png)
```
Comment (jubishop) [inline codelabs/custom_renderer/custom_renderer.md]: i'll want to change this alt-text based on how we decide to describe what this now looks like since it's not exactly the same as the default.

**Comment by jubishop** [comment]
should be done now, two outstanding questions.   ready for another round of review.

**Comment by jubishop** [comment]
also updated pull request message to match changes.

File: codelabs/custom_renderer/custom_renderer.md (hunk 3 above)
Comment (rachel-fenichel) [reply codelabs/custom_renderer/custom_renderer.md]: Discussed offline.

File: examples/custom-renderer-codelab/src/renderers/javascript.js (hunk 7 above)
Comment (rachel-fenichel) [reply examples/custom-renderer-codelab/src/renderers/javascript.js]: Yes, please add fileoverviews there too. At this point I consider them not required, but very nice to have for anyone who is exploring sample code.

File: codelabs/custom_renderer/custom_renderer.md (hunk 9 above)
Comment (jubishop) [reply codelabs/custom_renderer/custom_renderer.md]: done.

**Comment by jubishop** [comment]
ok. i tried to not get too wordy about it while removing the confusion week-ago justin had about geras vs base renderer.

**Review by rachel-fenichel — APPROVED** [review]
LGTM, and I'll merge once ci passes. Thanks!
